File: server/src/main/java/com/smartexpense/api/infrastructure/persistence/repository/ExpenseRepositoryImpl.java
package com.smartexpense.api.infrastructure.persistence.repository;

import com.smartexpense.api.domain.model.Expense;
import com.smartexpense.api.domain.repository.ExpenseRepository;
import com.smartexpense.api.infrastructure.persistence.entity.ExpenseEntity;
import com.smartexpense.api.infrastructure.persistence.mapper.ExpenseMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.stream.Collectors;

@Repository
@RequiredArgsConstructor
public class ExpenseRepositoryImpl implements ExpenseRepository {

    private final ExpenseJpaRepository jpaRepository;
    private final ExpenseMapper mapper;

    @Override
    public List<Expense> findAllByUserIdOrderByExpenseDateDesc(UUID userId) {
        return jpaRepository.findAllByUserIdOrderByExpenseDateDesc(userId)
                .stream()
                .map(mapper::toDomain)
                .collect(Collectors.toList());
    }

    @Override
    public Optional<Expense> findByIdAndUserId(UUID id, UUID userId) {
        return jpaRepository.findByIdAndUserId(id, userId)
                .map(mapper::toDomain);
    }

    @Override
    public Expense save(Expense expense) {
        ExpenseEntity entity = mapper.toEntity(expense);
        return mapper.toDomain(jpaRepository.save(entity));
    }

    @Override
    public void deleteByIdAndUserId(UUID id, UUID userId) {
        jpaRepository.deleteByIdAndUserId(id, userId);
    }
}

File: server/src/main/kotlin/com/smartexpense/api/infrastructure/persistence/repository/ExpenseRepositoryImpl.kt
package com.smartexpense.api.infrastructure.persistence.repository

import com.smartexpense.api.domain.model.Expense
import com.smartexpense.api.domain.repository.ExpenseRepository
import com.smartexpense.api.infrastructure.persistence.mapper.ExpenseMapper
import org.springframework.stereotype.Repository
import java.util.UUID

@Repository
class ExpenseRepositoryImpl(
    private val jpaRepository: ExpenseJpaRepository,
    private val mapper: ExpenseMapper
) : ExpenseRepository {

    override fun findAllByUserIdOrderByExpenseDateDesc(userId: UUID): List<Expense> {
        return jpaRepository.findAllByUserIdOrderByExpenseDateDesc(userId)
            .map { mapper.toDomain(it) }
    }

    override fun findByIdAndUserId(id: UUID, userId: UUID): Expense? {
        return jpaRepository.findByIdAndUserId(id, userId)?.let { mapper.toDomain(it) }
    }

    override fun save(expense: Expense): Expense {
        return mapper.toDomain(jpaRepository.save(mapper.toEntity(expense)))
    }

    override fun deleteByIdAndUserId(id: UUID, userId: UUID) {
        jpaRepository.deleteByIdAndUserId(id, userId)
    }
}
